File: src/document.ts
/**
 * Internal representation shared by both entry points.
 *
 * Both `convertJsonValue` and `convertJsonText` normalise their input into this
 * form so that a single renderer produces the Output Document. A number keeps
 * its Numeric Lexeme as text: the serialized path stores the original token, the
 * parsed path stores the caller's value formatted by the host.
 */
export type DocNode =
  | { readonly kind: "null" }
  | { readonly kind: "boolean"; readonly value: boolean }
  | { readonly kind: "number"; readonly lexeme: string }
  | { readonly kind: "string"; readonly value: string }
  | { readonly kind: "array"; readonly items: readonly DocNode[] }
  | { readonly kind: "object"; readonly entries: readonly (readonly [string, DocNode])[] };

export const NULL_NODE: DocNode = { kind: "null" };
export const TRUE_NODE: DocNode = { kind: "boolean", value: true };
export const FALSE_NODE: DocNode = { kind: "boolean", value: false };

export function isContainer(node: DocNode): boolean {
  return node.kind === "array" || node.kind === "object";
}

export function isEmptyContainer(node: DocNode): boolean {
  return (
    (node.kind === "array" && node.items.length === 0) ||
    (node.kind === "object" && node.entries.length === 0)
  );
}

/** A value shown inline (as a paragraph or list-item scalar) rather than as its own structure. */
export function rendersInline(node: DocNode): boolean {
  return !isContainer(node) || isEmptyContainer(node);
}

/** Zero-padded four-digit hex for a `\uXXXX` escape. */
export function hex4(code: number): string {
  return code.toString(16).padStart(4, "0");
}

/** RFC 6901 pointer segment escaping. */
export function pointerSegment(name: string): string {
  return name.replace(/~/g, "~0").replace(/\//g, "~1");
}

export function childPointer(pointer: string, name: string): string {
  return `${pointer}/${pointerSegment(name)}`;
}

File: go/document.go
package jsontomd

import (
	"fmt"
	"strings"
)

// Internal representation shared by both entry points, mirroring the
// TypeScript DocNode. A number keeps its Numeric Lexeme as text; an object
// keeps its members as an ordered slice — a Go map would lose Encounter
// Order. String values and member names are carried as WTF-8 so that lone
// UTF-16 surrogates parsed from \uXXXX escapes survive to render as visible
// escapes.
type nodeKind uint8

const (
	kindNull nodeKind = iota
	kindBool
	kindNumber
	kindString
	kindArray
	kindObject
)

type member struct {
	name  string
	value node
}

type node struct {
	kind    nodeKind
	boolVal bool
	lexeme  string // kindNumber: exact source spelling
	str     string // kindString: WTF-8 value
	items   []node
	entries []member
}

var (
	nullNode  = node{kind: kindNull}
	trueNode  = node{kind: kindBool, boolVal: true}
	falseNode = node{kind: kindBool}
)

func isContainer(n node) bool {
	return n.kind == kindArray || n.kind == kindObject
}

func isEmptyContainer(n node) bool {
	return (n.kind == kindArray && len(n.items) == 0) ||
		(n.kind == kindObject && len(n.entries) == 0)
}

// rendersInline reports a value shown inline (as a paragraph or list-item
// scalar) rather than as its own structure.
func rendersInline(n node) bool {
	return !isContainer(n) || isEmptyContainer(n)
}

// hex4 is the zero-padded four-digit hex used by \uXXXX escapes.
func hex4(code rune) string {
	return fmt.Sprintf("%04x", code)
}

// pointerSegment applies RFC 6901 escaping to one path segment.
func pointerSegment(name string) string {
	name = strings.ReplaceAll(name, "~", "~0")
	return strings.ReplaceAll(name, "/", "~1")
}

func childPointer(pointer, name string) string {
	return pointer + "/" + pointerSegment(name)
}
